pkg/builder/searcherclient/searcher_client.go

```diff
@@ -0,0 +1,226 @@
+	Count          int64 `json:"count"`
+	MinPriorityFee int64 `json:"MinPriorityFee"`
+	MaxPriorityFee int64 `json:"MaxPriorityFee"`
+}
+
+type Metadata struct {
+	Builder            string      `json:"builder"`
+	Number             int64       `json:"number"`
+	BlockHash          string      `json:"blockHash"`
+	Timestamp          string      `json:"timestamp"`
+	BaseFee            uint32      `json:"baseFee"`
+	Transactions       Transaction `json:"standard_transactions"`
+	ClientTransactions []string    `json:"personal_transactions,omitempty"`
+	SentTimestamp      time.Time   `json:"sent_timestamp"` // Timestamp of block sent to the searcher
+	RecTimestamp       time.Time   `json:"rec_timestamp"`  // Timestamp of block received by the builder instance
+}
+
+type SuperPayload struct {
+	InternalMetadata Metadata
+	SearcherTxns     map[string][]string
+}
+
+type searcherClient struct {
+	mu                   sync.Mutex
+	searchers            map[string]Searcher
+	disconnected         map[string]string
+	inclusionProofActive bool
+	logger               *slog.Logger
+}
+
+func NewSearcherClient(logger *slog.Logger, inclusionProof bool) SearcherClient {
+	return &searcherClient{
+		searchers:            make(map[string]Searcher),
+		logger:               logger,
+		inclusionProofActive: inclusionProof,
+	}
+}
+
+func (s *searcherClient) AddSearcher(searcher Searcher) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.searchers[searcher.ID()] = searcher
+	delete(s.disconnected, searcher.ID())
+}
+
+func (s *searcherClient) RemoveSearcher(id string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	searcher, found := s.searchers[id]
+	if !found {
+		return
+	}
+	err := searcher.Close()
+	if err != nil {
+		s.logger.Error("failed to close searcher", "err", err)
+	}
+	delete(s.searchers, searcher.ID())
+}
+
+func (s *searcherClient) Disconnected(id, reason string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.disconnected[id] = reason
+	delete(s.searchers, id)
+}
+
+func (s *searcherClient) SubmitBlock(
+	ctx context.Context,
+	payload *capella.SubmitBlockRequest,
+) error {
+	if payload == nil || len(payload.ExecutionPayload.Transactions) == 0 {
+		return nil
+	}
+
+	minTipTxn, maxTipTxn := big.NewInt(0), big.NewInt(0)
+	ts := time.Unix(int64(payload.ExecutionPayload.Timestamp), 0).Format(time.RFC1123)
+	baseFee := binary.LittleEndian.Uint32(payload.ExecutionPayload.BaseFeePerGas[:])
+
+	blockMetadata := SuperPayload{
+		InternalMetadata: Metadata{
+			RecTimestamp: time.Now(),
+			Builder:      payload.Message.BuilderPubkey.String(),
+			Number:       int64(payload.ExecutionPayload.BlockNumber),
+			BlockHash:    payload.Message.BlockHash.String(),
+			Timestamp:    ts,
+			BaseFee:      baseFee,
+			Transactions: Transaction{
+				Count: int64(len(payload.ExecutionPayload.Transactions)),
+			},
+		},
+	}
+	if s.inclusionProofActive {
+		blockMetadata.SearcherTxns = make(map[string][]string)
+	}
+
+	for idx, btxn := range payload.ExecutionPayload.Transactions {
```

Comment:
We should probably move these computations up a level, to avoid having to repeat them for all searchers.